GH-542 Adding Podspec for Cordova library

bhariharan opened the pull request:
<!--
Please make sure the checklist boxes are all checked before submitting the PR. The checklist is intended as a quick reference, for complete details please see our Contributor Guidelines:

http://cordova.apache.org/contribute/contribute_guidelines.html

Thanks!
-->

### Platforms affected
`cordova-ios`

### Motivation and Context
- Adds the ability to fetch `Cordova` through `Cocoapods`.
- closes #542

### Description
- Adds a `podspec` file that allows consumers to fetch the `Cordova` library through `Cocoapods`.

### Testing
- I have tested this by pulling this through `Cocoapods` in my own app and running it.

### Checklist

- [x] I've run the tests to see all new and existing tests pass
- [x] I added automated test coverage as appropriate for this change
- [x] Commit is prefixed with `(platform)` if this change only applies to one platform (e.g. `(android)`)
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))
- [x] I've updated the documentation if necessary

dpogue commented on Cordova.podspec:
```diff
@@ -0,0 +1,32 @@
+Pod::Spec.new do |s|
+  s.name         = "Cordova"
+  s.version      = "5.1.0"
```
Would it be possible to read this from the package.json file? We have too many copies of the version numbers in too many places, and our automated tool needs to try to keep track of all of them... if possible, it would be best to read the version from an existing spot that's guaranteed to be accurate.

janpio replied:
(If not, we probably have to add it to https://github.com/apache/cordova-coho/blob/master/docs/platforms-release-process.md#create-release-branch or better include it in https://github.com/apache/cordova-coho/blob/80f5dee833e095dc68de69434998b2e0f2575595/src/versionutil.js#L108-L109)

bhariharan replied:
It can't be read from `package.json` AFAIK, but it can be added to `versionutil.js`.

dpogue replied:
It looks like parsing package.json might be possible, based on what I see in the Podspec for React Native: https://github.com/facebook/react-native/blob/d0d8c8b651af6a2883307c777a1dfb753f046f7c/React.podspec#L7-L10

This would be preferable, since it's one less spot that could get missed during version bumps. The version in package.json **must** be correct, so it's safest to use that.

bhariharan replied:
Cool - I'll give it a shot. I'll try to get to these changes and submit them by this weekend.

dpogue commented on Cordova.podspec:
```diff
@@ -0,0 +1,32 @@
+Pod::Spec.new do |s|
+  s.name         = "Cordova"
+  s.version      = "5.1.0"
+  s.summary      = "Cordova for iOS"
```
I think this should say "Apache Cordova for iOS"

bhariharan replied:
Will do.

dpogue commented on Cordova.podspec:
```diff
@@ -0,0 +1,32 @@
+  s.author       = { "Bharath Hariharan" => "bhariharan@salesforce.com" }
+  s.platform     = :ios, "11.0"
+  s.source       = { :git => "https://github.com/apache/cordova-ios.git",
+                     :tag => "{s.version}",
```
I believe the convention for released tags is `rel/<version>`

bhariharan replied:
Will do.

dpogue commented on Cordova.podspec:
```diff
@@ -0,0 +1,32 @@
+  s.homepage     = "https://github.com/apache/cordova-ios"
+  s.license      = { :type => "Apache 2.0", :file => "LICENSE" }
+  s.author       = { "Bharath Hariharan" => "bhariharan@salesforce.com" }
+  s.platform     = :ios, "11.0"
```
I think the minimum iOS version currently supported is 10.0, not 11.0

bhariharan replied:
Will do.

dpogue commented on Cordova.podspec:
```diff
@@ -0,0 +1,32 @@
+  s.default_subspec  = 'Cordova'
+  s.subspec 'Cordova' do |cordova|
+      cordova.source_files = 'CordovaLib/Classes/**/*.{h,m}', 'CordovaLib/Cordova/Cordova.h'
+      cordova.public_header_files = 'CordovaLib/Classes/Public/CDV.h', 'CordovaLib/Classes/Public/CDVAppDelegate.h', 'CordovaLib/Classes/Public/CDVAvailability.h', 'CordovaLib/Classes/Public/CDVAvailabilityDeprecated.h', 'CordovaLib/Classes/Public/CDVCommandDelegate.h', 'CordovaLib/Classes/Public/CDVCommandDelegateImpl.h', 'CordovaLib/Classes/Public/CDVCommandQueue.h', 'CordovaLib/Classes/Public/ ...
```
Can we glob these headers instead of manually maintaining them?

bhariharan replied:
Yes - I have a magic script for this that updates it when the framework target is built. I can contribute that too.

janpio commented on Cordova.podspec:
```diff
@@ -0,0 +1,32 @@
+  s.summary      = "Cordova for iOS"
+  s.homepage     = "https://github.com/apache/cordova-ios"
+  s.license      = { :type => "Apache 2.0", :file => "LICENSE" }
+  s.author       = { "Bharath Hariharan" => "bhariharan@salesforce.com" }
```
Is this correct? Does this refer to who created the podspec or who "wrote"/owns the software?

bhariharan replied:
This refers to the original author.

jcesarmobile replied:
It should be "Apache Software Foundation"

jcesarmobile commented:
Before we do this, we should check if we are allowed to do this, not sure if Apache will allow us to distribute through CocoaPods.

BTW, some people has been publishing it already https://cocoapods.org/pods/Cordova, so we probably will need permissions from them, cc @shazron

janpio commented:
Afaik Apache doesn't really care about distribution method, as long as you _also_ create normal binaries that are kept in SVN as the base of truth.

bhariharan commented on Cordova.podspec:
```diff
@@ -0,0 +1,35 @@
+require "json"
+
+packageJson = JSON.parse(File.read(File.join(__dir__, "package.json")))
```
Updated to automatically pull in the version number from `package.json`.

bhariharan commented on Cordova.podspec:
```diff
@@ -0,0 +1,35 @@
+
+Pod::Spec.new do |s|
+  s.name         = "Cordova"
+  s.version      = relVersion
```
Version is auto-populated from above, i.e. `package.json`.

bhariharan commented on CordovaLib/CordovaLib.xcodeproj/project.pbxproj:
```diff
@@ -516,6 +517,26 @@
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "\"${PROJECT_DIR}/../update_podspec.sh\" -s cordova\n";
```
Added the shell script that updates podspec every time the library project is built. No more manual steps required to update the list of public headers in the podspec.

bhariharan commented on update_podspec.sh:
```diff
@@ -0,0 +1,70 @@
+set -e
```
Magic script to update podspec headers every time files are added/deleted in the library project.

bhariharan commented:
@erisu @jcesarmobile Could you help get this merged, please? All the feedback has been addressed and the podspec generation is completely automated now (no more manual update steps).

bhariharan commented:
@jcesarmobile @dpogue @janpio @erisu  Could you please review this and add it to the list for `6.0.0`? I have addressed all the feedback and verified everything works as expected. Happy to answer any questions you may have. Thanks!

brodybits commented:
@bhariharan I just pushed the milestone to 6.0.0 which we are now targeting on the master branch. I think many maintainers are taking a much deserved holiday break. I recommend that you ping us on Slack or on the mailing list in the new year, contact info is linked in the footer of cordova.io or cordova.apache.org.

bhariharan commented:
Thanks, @brodybits! Happy holidays!

jcesarmobile commented:
As I said, somebody already published it in CocoaPods so we don’t have permissions to publish. While it’s true that it can also be installed from GitHub, I find this mostly useless as most people will want the official release in CocoaPods.
I think this shouldn’t be merged until we get permissions to publish.


And please please please, stop pinging me.

brodybits commented:
I am with @jcesarmobile to not merge this before we obtain permission to publish updates to Cordova on Cocoapods. I do have another concern that I just added in <https://github.com/apache/cordova-ios/issues/542#issuecomment-569167351>.

I would like to apologize for all of the delays that this PR has gone through. I am sure that this proposal would help others in the user community and am grateful for your efforts on this.

I would like to *highly* recommend that you followup with us on Slack or on the mailing list, please follow the links in the footer of cordova.io or cordova.apache.org.

shortstuffsushi commented:
Hi there.

I'm one of the former maintainers of this podspec as well as a few of the plugins. I was added as a contributor by the original author at the time, and added several people I was working with at the time as well, since there wasn't interest from any Apache folks in maintaining it at the time.

I've been in contacted with @erisu recently who got me interested in this PR. I'm no longer using Cordova, so I was looking to hand this over (hasn't been updated for a couple years now..), so I've added him as an author for now. You're all welcome to add / remove maintainers as you please as this point. If you need anything from me, I'll be around - but I'm not doing any Cordova development these days, or any mobile at all lately, so I may not be able to speak too much to some of the specific questions.

shortstuffsushi commented on Cordova.podspec:
```diff
@@ -0,0 +1,35 @@
+      cordova.source_files = 'CordovaLib/Classes/**/*.{h,m}', 'CordovaLib/Cordova/Cordova.h'
+      cordova.public_header_files = 'CordovaLib/Classes/Public/CDV.h', 'CordovaLib/Classes/Public/CDVAppDelegate.h', 'CordovaLib/Classes/Public/CDVAvailability.h', 'CordovaLib/Classes/Public/CDVAvailabilityDeprecated.h', 'CordovaLib/Classes/Public/CDVCommandDelegate.h', 'CordovaLib/Classes/Public/CDVCommandDelegateImpl.h', 'CordovaLib/Classes/Public/CDVCommandQueue.h', 'CordovaLib/Classes/Public/ ...
+      cordova.prefix_header_contents = ''
+      cordova.requires_arc = true
```
As mentioned in the bigger comment, I'm a bit out of touch here, but there appear to be a few items missing from this subspec that appear in the published version, namely:

 * preserve_paths - for some of the assets
 * frameworks - for some of the required Core frameworks

bhariharan commented:
@erisu I will update this PR and address the outstanding feedback once your PR for removing `UIWebView` goes in. I imagine a bunch of classes will need to be updated here once that gets merged.

bhariharan commented:
Alright, I have brought in the changes from `master`, and verified that the script works as expected. The podspec has been updated as well. I have verified it's working as expected using our plugin (see [here](https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin/blob/master/plugin.xml#L106)). Our plugin uses the podspec from our fork of `Cordova` currently - I'll switch this to the upstream repo once this gets merged. Let me know if there are any questions/outstanding issues I should address.

bhariharan commented:
Could this be merged?

erisu commented on Cordova.podspec:
```diff
@@ -0,0 +1,37 @@
+  s.default_subspec  = 'Cordova'
+  s.subspec 'Cordova' do |cordova|
+      cordova.source_files = 'CordovaLib/Classes/**/*.{h,m}', 'CordovaLib/Cordova/Cordova.h'
+      cordova.public_header_files = 'CordovaLib/Classes/Public/CDV.h', 'CordovaLib/Classes/Public/CDVAppDelegate.h', 'CordovaLib/Classes/Public/CDVAvailability.h', 'CordovaLib/Classes/Public/CDVAvailabilityDeprecated.h', 'CordovaLib/Classes/Public/CDVCommandDelegate.h', 'CordovaLib/Classes/Public/CDVCommandDelegateImpl.h', 'CordovaLib/Classes/Public/CDVCommandQueue.h', 'CordovaLib/Classes/Public/ ...
```
Can the `public_header_files` property also be written in the same pattern like `source_files`?

```
cordova.public_header_files = 'CordovaLib/Classes/Public/**/*.{h}', 'CordovaLib/Cordova/Cordova.h'
```

I just noticed that this list is manually maintained and that the list is now not matching what is in master.
* `CDVUserAgentUtil.h` had been removed
* `CDVURLSchemeHandler.h` has been added

bhariharan replied:
@erisu No - public headers have to be explicitly listed. However, I added a script to automatically update this list every time a change is made. This list is not manually maintained - it's updated whenever `CordovaLib` is built in `Xcode`. It was behind only because this branch is behind `master`. I updated it to merge the latest from `master`, and the update podspec is in [this commit](https://github.com/apache/cordova-ios/pull/543/commits/8b9399e15f944d8f6b17a0c28b3a7273ec6b6f84). Basically, it will be updated automatically as part of the build process once this pull request is merged to `master`.
